### DISCUSSION TITLE
Does gomplate support .env file as data source?

### DISCUSSION THREAD

**Discussion by kundeng** [hairyhenderson/gomplate#2244] [Q&A]
New and learning!

**Comment by kundeng** [comment]
ok, I did this and I think it's working but it does removed all the comments when I try to use gomplate to merge .env files.  What would be a better way?  Also the merged env doesn't seem to keep the ordering of either original env either.  Keys seems to have been sorted. 


global.env:
```
DOMAIN=vlab.local
TimeZone=America/Detroit
```
.env.tmpl: 
```
# Stack-specific environment variables
DOMAIN=${DOMAIN}
API_URL=${API_URL}
FIELD=custom_value
```

# merging logic: 

```
{{- /* Retrieve the 'global' and 'stack' datasources using their aliases */ -}}
{{- $global := ds "global" -}}
{{- $stack := ds "stack" -}}

{{- /* Merge the two datasources, with stack-specific variables overriding global ones */ -}}
{{- $merged := coll.Merge $global $stack -}}

{{- /* Iterate over the merged key-value pairs and output them in .env format */ -}}
{{- range $key, $value := $merged }}
{{ $key }}={{ $value }}
{{- end }}
```

It results in:
```

API_URL=
DOMAIN=vlab.local
FIELD=custom_value
TimeZone=America/Detroit

```

**Reply by hairyhenderson** [reply]
> Keys seems to have been sorted.

Correct - that's the intended behaviour. The reason is that gomplate parses your data into a map structure, where the items are no longer ordered. In fact, by default, iterating through a map like this has unstable ordering (read about it [here](https://go.dev/blog/maps#iteration-order) for more detail).

So, for a slightly more user-friendly experience, gomplate will always iterate through maps in lexical order.

If you absolutely must have a defined order for your keys, you could define a list of all keys, then iterate through that and access your merged map that way instead. I'll leave that as an exercise for you, if you really need this.
